[SPARK-23619][DOCS] Add output description for some generator expressions / functions

jashgala opened the pull request:
## What changes were proposed in this pull request?

This PR addresses SPARK-23619: https://issues.apache.org/jira/browse/SPARK-23619

It adds additional comments indicating the default column names for the `explode` and `posexplode` 
functions in Spark-SQL.

Functions for which comments have been updated so far:
* stack
* inline
* explode
* posexplode
* explode_outer
* posexplode_outer

## How was this patch tested?

This is just a change in the comments. The package builds and tests successfullly after the change.

jashgala committed:
[SPARK-23619][SQL] Added default column names for explode & posexplode functions

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

maropu commented:
You need to update `ExpressionDescription` if you want to update the docs, e.g.,
https://github.com/apache/spark/blob/d0443a74d185ec72b747fa39994fa9a40ce974cf/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala#L355
https://spark.apache.org/docs/2.4.0/api/sql/index.html#explode

maropu commented:
Also, you need to update the other functions for the same fix. e.g., `explode_outer`, ...

jashgala committed:
Added more comments for explode_outer and posexplode_outer

jashgala committed:
Added default column names in ExpressionDescription for explode and posexplode

jashgala commented:
@maropu 

I've fixed the comments, added comments for `explode_outer` and `posexplode_outer` and also added the ExpressionDescriptions.

Please let me know if there is anything else I might've missed.

maropu commented:
Thanks. Could you cover all the related document fixes in this pr? How about `Inline` and `Stack`? They have default column names, right?

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/functions.scala:
```diff
@@ -3305,6 +3305,8 @@ object functions {
   /**
    * Creates a new row for each element in the given array or map column.
+   * Uses the default column name `col` for elements in the array and
+   * `key` and `value` for elements in the map unless specified otherwise.
```
Let's also update Python doc and R doc while we're here. Take a look for functions.py and functions.R

maropu replied:
oh... I forgot the viewpoint...thanks.

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #102194 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102194/testReport)** for PR 23748 at commit [`4ed035b`](https://github.com/apache/spark/commit/4ed035b302e1f15d7331458e6bd14ece3073564d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7811/
Test PASSed.

SparkQA commented:
**[Test build #102195 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102195/testReport)** for PR 23748 at commit [`4ed035b`](https://github.com/apache/spark/commit/4ed035b302e1f15d7331458e6bd14ece3073564d).

SparkQA commented:
**[Test build #102194 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102194/testReport)** for PR 23748 at commit [`4ed035b`](https://github.com/apache/spark/commit/4ed035b302e1f15d7331458e6bd14ece3073564d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102194/
Test FAILed.

SparkQA commented:
**[Test build #102195 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102195/testReport)** for PR 23748 at commit [`4ed035b`](https://github.com/apache/spark/commit/4ed035b302e1f15d7331458e6bd14ece3073564d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102195/
Test FAILed.

jashgala commented:
@maropu @HyukjinKwon 

While looking at the default column names used by inline and stack, I found that inline uses col1, col2, etc. (i.e. 1-indexed columns), while stack uses col0, col1, col2, etc. (i.e. 0-indexed columns).

```
scala> spark.sql("SELECT stack(2, 1, 2, 3)").show
+----+----+
|col0|col1|
+----+----+
|   1|   2|
|   3|null|
+----+----+


scala>  spark.sql("SELECT inline_outer(array(struct(1, 'a'), struct(2, 'b')))").show
+----+----+
|col1|col2|
+----+----+
|   1|   a|
|   2|   b|
+----+----+
```

This feels like an issue with consistency. Is there a reason why this column naming convention is kept different?

maropu commented:
ah, yes. I think it'd better to fix this, too, in a separate pr.
Probably, I think zero-based indexing is more natural. For example, auto-generated column names in CSV read is baed on zero-based;
```
scala> spark.read.csv("test.csv").show
+---+---+---+                                                                   
|_c0|_c1|_c2|
+---+---+---+
|  1|  2|  3|
+---+---+---+
```

Could you check if we have other expressions having one-based indexing?

jashgala commented:
Have created a JIRA for tracking the consistency fix: https://issues.apache.org/jira/browse/SPARK-26879

Will update the same other expressions having one-based indexing on this JIRA itself, so that it can continue as a separate investigation.

jashgala committed:
Added default column names for inline and stack in comments

SparkQA commented:
**[Test build #102622 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102622/testReport)** for PR 23748 at commit [`e5751ed`](https://github.com/apache/spark/commit/e5751ed9ce397b8d634482ea4f41a383a55e271e).

SparkQA commented:
**[Test build #102622 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102622/testReport)** for PR 23748 at commit [`e5751ed`](https://github.com/apache/spark/commit/e5751ed9ce397b8d634482ea4f41a383a55e271e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102622/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8173/
Test PASSed.

jashgala committed:
Added scala style line-size exception to prevent failure due to comment

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8174/
Test PASSed.

SparkQA commented:
**[Test build #102623 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102623/testReport)** for PR 23748 at commit [`80c0649`](https://github.com/apache/spark/commit/80c0649f35acff057bcea51450606260d78caa46).

SparkQA commented:
**[Test build #102623 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102623/testReport)** for PR 23748 at commit [`80c0649`](https://github.com/apache/spark/commit/80c0649f35acff057bcea51450606260d78caa46).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102623/
Test FAILed.

jashgala committed:
Fixed typo in scala style exceptions in comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8175/
Test PASSed.

SparkQA commented:
**[Test build #102624 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102624/testReport)** for PR 23748 at commit [`d029d3c`](https://github.com/apache/spark/commit/d029d3c7f0883372b76adad531c6a25abea0d59a).

jashgala committed:
Added default column names in comments for explode, etc. functions in R

jashgala committed:
Added default column names in comments for explode, etc. functions in Python

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8176/
Test PASSed.

SparkQA commented:
**[Test build #102625 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102625/testReport)** for PR 23748 at commit [`d0dea9c`](https://github.com/apache/spark/commit/d0dea9cc79b01bf78b1c45bdec3012890bc90718).

jashgala commented:
@maropu I've added comments for Inline and Stack in the `generators.scala` file. I didn't see these in the functions.* files though.
@HyukjinKwon Have added comments in Python and R functions files as well as per the review comments above.

I see that there are a lot of functions (e.g. mathematical functions, etc.) in the `functions.*` files. Is there value in adding the column names in the comments for all of them or do should we make this change specifically for the functions that are specifically used to generate columns/rows (e.g. explode, etc.)?

HyukjinKwon commented:
I haven;t checked super closely but looks fine.

SparkQA commented:
**[Test build #102625 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102625/testReport)** for PR 23748 at commit [`d0dea9c`](https://github.com/apache/spark/commit/d0dea9cc79b01bf78b1c45bdec3012890bc90718).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #102624 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102624/testReport)** for PR 23748 at commit [`d029d3c`](https://github.com/apache/spark/commit/d029d3c7f0883372b76adad531c6a25abea0d59a).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102625/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102624/
Test FAILed.

dilipbiswal commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8179/
Test PASSed.

SparkQA commented:
**[Test build #102631 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102631/testReport)** for PR 23748 at commit [`d0dea9c`](https://github.com/apache/spark/commit/d0dea9cc79b01bf78b1c45bdec3012890bc90718).

maropu commented:
@jashgala Yea, I think we would be nice to fix all the same issues in this pr where possible.

SparkQA commented:
**[Test build #102631 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102631/testReport)** for PR 23748 at commit [`d0dea9c`](https://github.com/apache/spark/commit/d0dea9cc79b01bf78b1c45bdec3012890bc90718).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102631/
Test PASSed.

jashgala commented:
Cool... Will make the change and add more commits!

jashgala commented:
@maropu 
Just wanted to update... I'm going through each function, understanding it and then adding the comments... It's taking a bit of a while especially due to real life responsibilities, but I want to reassure I am working on it!

HyukjinKwon commented:
I took a quick look and looks fine. Let me leave it to @maropu since he's been actively reviewing this.

maropu commented:
Could you update the PR description; plz list up all the function you addressed in this PR?

jashgala commented:
So far I've addressed the following (also updated in PR description):
* stack
* inline
* explode
* posexplode
* explode_outer
* posexplode_outer

The others are still WIP

HyukjinKwon commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10237/
Test PASSed.

SparkQA commented:
**[Test build #104934 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104934/testReport)** for PR 23748 at commit [`d0dea9c`](https://github.com/apache/spark/commit/d0dea9cc79b01bf78b1c45bdec3012890bc90718).

SparkQA commented:
**[Test build #104934 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104934/testReport)** for PR 23748 at commit [`d0dea9c`](https://github.com/apache/spark/commit/d0dea9cc79b01bf78b1c45bdec3012890bc90718).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104934/
Test PASSed.

HyukjinKwon commented:
Merged to master.

jashgala commented:
Thanks.... I will fix the other documentation when I have some free time!
